### PR TITLE
Add feature to create user avatar thumbnail

### PR DESCRIPTION
**Pull request by laura-barluzzi**
### Proposed changes in this pull request

This PR creates a thumbnail out of the `user.avatar` image in order to display the `user.avatar` in the user dashboard (work in progress). Merging this would enable future use of `user.avatar` image across the platform with any desired size. Part of epic User Dashboard #1491. 

- I copied the existing function `resources/models.create_thumbnails` and pasted into `resources/utils/thumbnail.py`
- I changed the name of this copied-and-pasted function into `create_image_thumbnails`
- I set *3 parameters `instance, s3_image_field_name, size=(150, 150)`*. This enables us to use this function freely from any model with a S3 file field while also allowing to set the desired thumbnail size
- Given that `user.avatar` accepts only `image/png, image/jpg` types, this function currently works *only* for image files. This simply means that you call this function ***after*** checking if the file is an image. *(see follow-up action)
- I created a couple of private property functions so that we call `create_image_thumbnails` only when necessary (that is when the `_avatar_has_changed` and the `_avatar_is_custom`)
- In `accounts/profile.html` I changed the `` to `` and I made some minor adjustments to the CSS. Here you see the views of `accounts.profile.html` after the changes:

- I added tests in `resources/tests/test_utils.py` using patch and mocking `thumbnail.make()` in order to cover the cases when:
1. the passed `s3_image_field_name` name does not exist in the given `instance`
2. the `url` attribute of `s3_image_field_name` is empty.
- I added tests in `accounts/tests/test_models.py` to cover all new property functions

### When should this PR be merged

As soon is approved.

### Risks

Low.

### Follow-up actions

- Ideally expand this and combine with the current parallel work in `PR Outreachy PDF Form generation #1409` in order to achieve an easy-to-customize version that enables to use the thumbnail functionality from ***any*** model with a file stored in S3 and that covers even *non-image cases* (e.g. icons).

#### General

- **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description. 
  - [ ] Review 1
  - [ ] Review 2
- **Is the PR labeled correctly?** It should have the `migration` label if a new migration is added. 
  - [ ] Review 1
  - [ ] Review 2
- **Is the risk level assessment sufficient?** The risks section should contain all risks that might be introduced with the PR and which actions we need to take to mitigate these risks. Possible risks are database migrations, new libraries that need to be installed or changes to deployment scripts.
  - [ ] Review 1
  - [ ] Review 2

#### Functionality

- **Are all requirements met?** Compare implemented functionality with the requirements specification.
  - [ ] Review 1
  - [ ] Review 2
- **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled. 
  - [ ] Review 1
  - [ ] Review 2

#### Code

- **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
  - [ ] Review 1
  - [ ] Review 2
- **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests. 
  - [ ] Review 1
  - [ ] Review 2
- **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/). 
  - [ ] Review 1
  - [ ] Review 2
- **Is the code documented sufficiently?** Large and complex classes, functions or methods must be annotated with comments following our [code-style guidelines](https://devwiki.corp.cadasta.org/Contributing%20Style%20Guide#documentation-and-comments).
  - [ ] Review 1
  - [ ] Review 2
- **Has the scalability of this change been evaluated?**
  - [ ] Review 1
  - [ ] Review 2
- **Is there a maintenance plan in place?**
  - [ ] Review 1
  - [ ] Review 2

#### Tests

- **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components. 
  - [ ] Review 1
  - [ ] Review 2
- **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
  - [ ] Review 1
  - [ ] Review 2
- **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?
  - [ ] Review 1
  - [ ] Review 2

#### Security

- **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
  - [ ] Review 1
  - [ ] Review 2
- **Are all UI and API inputs run through forms or serializers?** 
  - [ ] Review 1
  - [ ] Review 2
- **Are all external inputs validated and sanitized appropriately?**
  - [ ] Review 1
  - [ ] Review 2
- **Does all branching logic have a default case?**
  - [ ] Review 1
  - [ ] Review 2
- **Does this solution handle outliers and edge cases gracefully?**
  - [ ] Review 1
  - [ ] Review 2
- **Are all external communications secured and restricted to SSL?**
  - [ ] Review 1
  - [ ] Review 2

#### Documentation

- **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes must be documented in the [Cadasta Platform Documentation](https://github.com/Cadasta/cadasta-docs).
  - [ ] Review 1
  - [ ] Review 2
- **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented in the [API docs](https://github.com/Cadasta/api-docs).
  - [ ] Review 1
  - [ ] Review 2
- **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki. 
  - [ ] Review 1
  - [ ] Review 2

### CODE REVIEW

File: cadasta/resources/tests/test_utils.py [hunk 1]
```diff
@@ -35,3 +38,16 @@ def test_make_rotated(self):
         thumb = thumbnail.make(image, (100, 100))
         assert thumb.size[0] == 100
         assert thumb.size[1] == 100
+
+    @patch.object(thumbnail, 'make')
+    def test_create_image_thumbnails_without_filefield_or_url(self, mock_make):
```
Comment (oliverroick) [inline cadasta/resources/tests/test_utils.py]: With this test, you only test `create_image_thumbnails` up to line 106. All following lines are only implicitly covered by [this test](https://github.com/Cadasta/cadasta-platform/blob/ae10b199c7b92b67ba5eb0e23c1316984f7175e2/cadasta/accounts/tests/test_forms.py#L232), for example.

We should add tests for the full functionality in `create_image_thumbnails` so we can verify it's working correctly in isolation of other model updates.

**Comment by laura-barluzzi** [comment]
@amplifi  and @oliverroick thank you for your comments. Before going onto the details, I'd like to ask 2 questions: 
1. _Why the existing Outreachy PDF Form generation #1409 is approved if it applies this same approach?_

I just want to point out that I used this approach because I sort of **copied it from that PR which is approved**. Further, there aren't comments on this regard in that long PR thread.

This can be misleading since it states that this approach is approved, so maybe that PR should not be approved? Or at least this issue should be mentioned in the comment thread? 

2. _Is it possible to just store the temporary files using S3? Can we just create a S3 folder called 'temp' and use it for these things? Is this a very-bad-idea? :)_ 

I get back with more details on a proposal on how to overcome all this.

**Comment by amplifi** [comment]
@laura-barluzzi Yes, it is confusing that #1409 is approved. It shouldn't be, will also require refactoring before merge. Its use of local disk as temp space isn't authorized. That feature was planned before the async pipeline work was commenced and before our current process of approvals for architecture decisions. The same concerns should have been identified during the review process for that PR.

**Comment by laura-barluzzi** [comment]
@amplifi ok thank you very much for the fast clarification. These stuff happens, and we should now focus on how to find a solution.

I changed my previous comment since I realized I proposed something that didn't make sense :) 

Could you please address my 2 questions to see a solution looking forward? Cheers.

**Comment by laura-barluzzi** [comment]
@amplifi  and @oliverroick Chandra and I decided not to work on this since my internship ends in 3 weeks and I have also other tasks.

This said, I'd like to point out the importance to clear-up the security issue at the dev team level. This mainly because:
- Currently the platform uses the thumbnail approach that is not approved; 
- There aren't clear efforts towards fixing the previous point (e.g. beyond comments from my finished PRs, I never heard people talking about "fixing the resource thumbnail" which is in production)
- Developers work on PRs simulating the working-in-production approach (thumbnails for resources) but then get blocked. This is ineffective since the information arrives at the end and is also contrasting with point one.

I believe that this highlights the importance for a "user file security topic" conversation with all devs. For example, I believe these questions should be answered and agreed upon in the dev team: 
- *what do we do with files that are uploaded from users but we need to process? e.g. thumbnail now, but who knows in the future?*
- *should we delete the existing approach until we don't find a solution?* (e.g. delete the thumbnails of resources)
- *should we add this notice in the contribution guidance?* (e.g. notify all contributors to keep in mind not to copy this approach since it must be changed soon)
- *does everyone agree with this zero-risk taking approach?* (e.g. has this security change been agreed by everyone? is the opinion of few?) I personally heard different opinions.

I just hope this may help.

**Comment by amplifi** [comment]
@laura-barluzzi I understand the confusion around this approach, and our approved methodologies should have been communicated to you before commencing work. I'll attempt to clarify and provide some background information that might help going forward:

> Currently the platform uses the thumbnail approach that is not approved;
There aren't clear efforts towards fixing the previous point (e.g. beyond comments from my finished PRs, I never heard people talking about "fixing the resource thumbnail" which is in production)

The Cadasta development team [approved](https://github.com/Cadasta/decision-records/pull/7) a [decision record](https://github.com/Cadasta/decision-records/blob/master/0006_asynchronous_import_and_export.md) on 22 February to stop processing file imports and exports using synchronous processes, which includes no longer saving files to local server disk. This decision record specifically mentions the use cases of project data imports, project data exports, resource uploads, user avatar uploads, and thumbnail creation. The asynchronous framework corresponding to this DR has been completed, and refactoring the export process to utilize it is underway. Any new features should be developed under the same framework, which is why the PDF generator PR and this one should be re-worked, otherwise we'll be deploying code that we already know needs to be refactored. We also have an outstanding task to refactor search exports to use the same framework. Saving user uploads to server disk is an anti-pattern that we should actively seek to avoid.

> Developers work on PRs simulating the working-in-production approach (thumbnails for resources) but then get blocked. This is ineffective since the information arrives at the end and is also contrasting with point one.

This is an important point -- just because something has been done a certain way in the past does not mean it is a good way or should be continued. We're facing a significant amount of technical debt, so there are many aspects of the Cadasta Platform code that is intended to change.

> I believe that this highlights the importance for a "user file security topic" conversation with all devs.

We have held several development team conversations along these lines, including the month-long review period for the decision record above, in dev team meetings, and regarding other PRs that have needed re-work for similar reasons. But I agree it should be added to the contribution guidelines, or at very least contributors directed to reference the [decision record repo](https://github.com/Cadasta/decision-records) to ensure any new work is architected in line with our documented strategies. All Cadasta decision records are presented to the entire development team for discussion and review, and only merged into the repo after consensus and approval, so this is a solid starting point.

> For example, I believe these questions should be answered and agreed upon in the dev team:
> what do we do with files that are uploaded from users but we need to process? e.g. thumbnail now, but who knows in the future?

This is covered under the decision record.

> should we delete the existing approach until we don't find a solution? (e.g. delete the thumbnails of resources)

We have addressed this on a case-by-case basis for features implemented prior to the decision record above, as determined by the potential for negative impact. Critical risks have been reverted, while lower risk cases like thumbnailing are awaiting replacement under the async task banner.

> should we add this notice in the contribution guidance? (e.g. notify all contributors to keep in mind not to copy this approach since it must be changed soon)

Yes, I think this is a great idea.

> does everyone agree with this zero-risk taking approach? (e.g. has this security change been agreed by everyone? is the opinion of few?) I personally heard different opinions.

The decision record was approved by the development team. Once a decision record is approved, our process for any changes is to open a new PR for that decision record, distribute for review and discussion, and obtain approval for those changes. 

The approach outlined in the decision record is not zero-risk, but it follows best practice and greatly reduces the possibility of malicious file uploads and denial of service attacks. It also ensures compliance with our grant funding. Given the highly sensitive nature of the information stored in the Cadasta Platform, data security is not a responsibility we can take lightly.

I hope that puts things in context, but please feel free to reach out on Slack if you have any questions!

**Comment by laura-barluzzi** [comment]
I closed this PR so that is not perceived as work. I still believe that once the issues of security and async are established, this PR could be used as inspiration on how to generalize the thumbnail approach functionalities.

**Pull request closed without merging**